End to end test

wjwwood opened the pull request:
This pull request:

- moves `ros2launch_security` into a subfolder
- adds `ros2launch_security_example` which can serve as an example and an end-to-end test

This draft still because I need to:

- [x] push the changes for the automation of the test
- [x] investigate issues and work arounds when using with node namespaces

wjwwood committed:
move ros2launch_security into a subdirectory

Signed-off-by: William Woodall <william@osrfoundation.org>

wjwwood committed:
add example

Signed-off-by: William Woodall <william@osrfoundation.org>

wjwwood committed:
adding automated test for example

Signed-off-by: William Woodall <william@osrfoundation.org>

wjwwood committed:
adding test that uses ros2 launch --secure

Signed-off-by: William Woodall <william@osrfoundation.org>

wjwwood committed:
linters

Signed-off-by: William Woodall <william@osrfoundation.org>

wjwwood commented:
One thing I ran into today is that the integration of the `--secure` option for `ros2 launch` may need a similar integration with `launch_testing`, which does not use `ros2 launch` to run things. To work around this I just have a launch test that calls `ros2 launch` with the `--secure` option, but that's a pretty roundabout way to do it.

@gbiggs fyi

gbiggs commented:
I haven't got much experience with `launch_testing`, so I didn't know it doesn't use `launch`. I'd guess that the maintainer of this package is also not aware.

wjwwood commented:
> I haven't got much experience with launch_testing, so I didn't know it doesn't use launch. I'd guess that the maintainer of this package is also not aware.

For now I'll just ticket this somewhere. My workaround of running `ros2 launch` as part of the test is kind of ok, but it would be good to change in the future.

wjwwood commented:
@gbiggs I can't seem to request a review from @Arnatious, are you sure he has permissions on this repository?

gbiggs commented:
He's been invited, but the invite is still pending.

gbiggs commented on ros2launch_security_examples/test/test_fake_imu_launch_with_security.py:
```diff
@@ -0,0 +1,97 @@
+            (None, os.path.join(g_this_dir, 'expected_outputs', 'imu_sink')),
+            # TODO(wjwwood): right now, looking for the "Found security directory ..."
+            #   messages output anywhere, but it would be better to assert that
+            #   the outptu is cominig from the right processes, but that would
```
```suggestion
            #   the output is coming from the right processes, but that would
```

wjwwood replied:
ad2ecd7

gbiggs commented on ros2launch_security_examples/test/test_fake_imu_launch_with_security.py:
```diff
@@ -0,0 +1,97 @@
+            #   messages output anywhere, but it would be better to assert that
+            #   the outptu is cominig from the right processes, but that would
+            #   require figuring out how to integrate the `--secure` option for
+            #   'ros2 launch' into launch_testinig...
```
```suggestion
            #   'ros2 launch' into launch_testing...
```

wjwwood replied:
ad2ecd7

clalancette commented on ros2launch_security_examples/CMakeLists.txt:
```diff
@@ -0,0 +1,95 @@
+# nodl_export_node_description_file(nodl/imu_sink.nodl.xml)
+nodl_export_node_description_file(imu_sink.nodl.xml)
+
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  ament_lint_auto_find_test_dependencies()
+endif()
```
This is a duplicate of below.  Looks like a bad merge, maybe?

wjwwood replied:
You're right! I think it was left over from `ros2 pkg create`.

clalancette commented on ros2launch_security_examples/fake_imu.nodl.xml:
```diff
@@ -0,0 +1,6 @@
+    <topic name="imu" type="sensor_msgs/msg/Imu" role="publisher" />
+    <service name="reset_imu" type="example_interfaces/srv/Trigger" role="server" />
+  </node>
+</interface>
```
Nit: add a trailing newline.

wjwwood replied:
73675e6

clalancette commented on ros2launch_security_examples/imu_sink.nodl.xml:
```diff
@@ -0,0 +1,6 @@
+    <topic name="imu" type="sensor_msgs/msg/Imu" role="subscription" />
+    <service name="reset_imu" type="example_interfaces/srv/Trigger" role="client" />
+  </node>
+</interface>
```
Nit: add a trailing newline.

wjwwood replied:
73675e6

clalancette commented on ros2launch_security_examples/test/test_fake_imu_launch.py:
```diff
@@ -0,0 +1,89 @@
+                timeout=30,
+            )
+
+        # # TODO(hidmic): either make the underlying executables resilient to
+        # # interruptions close/during shutdown OR adapt the testing suite to
+        # # better cope with it.
+        # import time
+        # time.sleep(5)
```
What's the deal with this block?  Should we just remove it?

wjwwood replied:
I ran the tests a bunch of times this afternoon (in a loop) and I didn't see any issues. I suppose it could result in flaky tests, but honestly I don't know the root issue this is trying to fix. It was from here:

https://github.com/ros2/demos/blob/3af19c93c7abd9a83caca47de0780e941faeb22b/demo_nodes_cpp/test/test_executables_tutorial.py.in#L60-L64

I also tried interrupting the tests mid test, and early after the test starts, but I wasn't able to cause any unexpected issue like a hang or something.

wjwwood replied:
I'm tempted to just remove it.

wjwwood replied:
9901854

clalancette commented on ros2launch_security_examples/test/test_fake_imu_launch_with_security.py:
```diff
@@ -0,0 +1,97 @@
+        SetEnvironmentVariable('RCUTILS_CONSOLE_OUTPUT_FORMAT', '{message}')
+    )
+    launch_description.add_action(
+        ExecuteProcess(
+            cmd=[
+                'ros2', 'launch',
+                os.path.join(g_this_dir, '..', 'launch', 'fake_imu.launch.xml'),
+                '--secure',
+            ],
```
I think it would be good to put a comment here explaining why you have to do things this way and not use the builtin `IncludeLaunchDescription` as you do in the non-security version.

wjwwood replied:
This comment below it explains a bit, but I can put one here too:

https://github.com/osrf/ros2launch_security/pull/4/files#diff-71305fd43c410f758e5a1ece15052f54aa928231021fd86a401b2a5f4bff0309R72-R76

73675e6

clalancette commented on ros2launch_security_examples/test/test_fake_imu_launch_with_security.py:
```diff
@@ -0,0 +1,97 @@
+                ),
+                output_filter=output_filter,
+                timeout=10,
+                # This is needed because ros2 launch channels things to stdout.
```
Do you mean `stderr` here?

wjwwood replied:
No, actually I mean `stdout`. This test filter defaults to reading from `stderr`, I guess because the console logging from our applications default to outputting on `stderr`, but `ros2 launch` captures all of that and outputs it on `stdout`. So I had to explicitly tell this filter to check `stdout` instead or it didn't work.

clalancette replied:
Ah, I wasn't aware of that.  Thanks for the explanation.

clalancette reviewed (commented):
I've left a bunch of pretty minor comments.

wjwwood committed:
fix typos

Signed-off-by: William Woodall <william@osrfoundation.org>

wjwwood committed:
review feedback

Signed-off-by: William Woodall <william@osrfoundation.org>

wjwwood committed:
remove vestigial fix from test I copied from

Signed-off-by: William Woodall <william@osrfoundation.org>

aprotyas commented:
@wjwwood Can the `test_fake_imu_launch*` tests be flaky? Sometimes the test does not timeout in 10 seconds and is reported as a failure.

wjwwood commented:
@aprotyas maybe? I removed some logic that may have been there to reduce flaky behavior because I didn't understand it. I never experienced an issue with them though.

It could also be something unrelated to that logic I removed that makes it flaky.

wjwwood commented:
See this comment thread: https://github.com/osrf/ros2launch_security/pull/4#discussion_r661691680

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14874)](http://ci.ros2.org/job/ci_linux/14874/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9624)](http://ci.ros2.org/job/ci_linux-aarch64/9624/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12541)](http://ci.ros2.org/job/ci_osx/12541/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15033)](http://ci.ros2.org/job/ci_windows/15033/)

wjwwood commented:
The tests are failing in an unexpected way, I'm looking into it.

wjwwood commented:
I think I found the culprit change, I'll keep looking into what needs to change here or there: https://github.com/ros2/launch/pull/516#issuecomment-879445998

wjwwood commented:
I'm still waiting for https://github.com/ros2/launch_ros/pull/256#issuecomment-889568164.

wjwwood commented:
We're going to go forward without the fix from https://github.com/ros2/launch_ros/pull/256#issuecomment-889568164, because it only affects tests on Windows, is not a fix that will happen here (most likely will be contained to launch packages).
